x-pack/filebeat/input/http_endpoint/validate.go

```diff
@@ -52,3 +68,35 @@ func (v *apiValidator) ValidateHeader(r *http.Request) (int, error) {
 
 	return 0, nil
 }
+
+func (v *hmacValidator) Validate(r *http.Request) (int, error) {
+	if v.hmacHeader != "" && v.hmacToken == "" || v.hmacHeader == "" && v.hmacToken != "" {
+		return http.StatusMethodNotAllowed, fmt.Errorf("Both hmacToken and hmacHeader has to be set")
+	}
+
+	if v.hmacToken != "" && v.hmacHeader != "" {
+		if len(r.Header.Get(v.hmacHeader)) == 0 {
+			return http.StatusInternalServerError, fmt.Errorf("The HMAC signature in the configured request header is empty")
+		}
+		s := hmac.New(sha1.New, []byte(v.hmacToken))
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("Failed to read the request body: %v", err)
+		}
+		r.Body.Close()
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+
+		s.Write(b)
+		h := r.Header.Get(v.hmacHeader)
+		// If the header includes a prefix before the SHA-1 key, we need to only grab the signature after the prefix. Can also be 0
+		hWithPrefix := make([]byte, 20)
+		hex.Decode(hWithPrefix, []byte(h[len(v.hmacPrefix):]))
+
+		if !hmac.Equal(s.Sum(nil), hWithPrefix) {
+			return http.StatusUnauthorized, errIncorrectHmac
+		}
+
+	}
+
+	return 0, nil
+}
```

Comment:
hm, this is doing more than just validating the contents. This looks like an intermediate data processing level. Maybe this should be some middleware layer that wraps the body. The custom reader would stream all writes from body to the hmac 'validator'. If we hit EOF, the reader does the final check and fails if there was a failure. In case the wrapped handler did not consume the body, the middleware handler would read until EOF (or error) in order to enforce the HMAC check.

Using layers we would create a processing chain like: `bodyLimit -> hmac -> header validation -> body validation`. The `bodyLimit` would wrap the body and fail in case the body (or contents-length) is bigger then the configured limits.

Are there packages/projects providing authentication support as middleware, we can use?